godot-cell/tests/mock.rs

```diff
@@ -433,5 +443,43 @@ fn calls_parallel_many_parallel() {
     unsafe {
         assert!(get_int(instance_id) >= min_expected);
         assert!(get_int(instance_id) <= max_expected);
+
+        #[cfg(feature = "threads")]
+        assert!(get_int(instance_id) == max_expected);
     }
 }
+
+/// Rwborrow the same cell on multible threads.
+///
+/// This verifies that threads don't hang if a reborrow occours inside of them, i.e. immutable borrow followed
+/// by a mutable borrow. Threads should only block if no local borrow exists but an other thread already has a
+/// mutable borrow or other threads already hold immutable borrows but the local thread requests it's first and only mutable borrow.
+#[test]
+fn non_blocking_reborrow() {
+    use std::thread;
+    let instance_id = MyClass::init();
+
+    let thread_a = thread::spawn(move || unsafe {
+        call_immut_method(instance_id, MyClass::immut_calls_mut_directly).unwrap();
+    });
+
+    let thread_b = thread::spawn(move || unsafe {
+        call_immut_method(instance_id, MyClass::immut_calls_mut_directly).unwrap();
+    });
+
+    println!("thread_main: {:?}", thread::current().id());
+    println!("thread_a: {:?}", thread_a.thread().id());
+    println!("thread_b: {:?}", thread_b.thread().id());
+
+    let panic_a = thread_a.join().err();
+
+    assert!(panic_a.is_some());
+
+    println!("{:?} -----", panic_a.unwrap());
+
+    let panic_b = thread_b.join().err();
+
+    assert!(panic_b.is_some());
+
+    println!("{:?} +++++", panic_b.unwrap());
```

Comment:
Maybe instead of printing, compare it to a known output? No one is going to check manually how the panic is worded 🙂

`Debug` representations are technically not stable, but if they change rarely, the maintenance effort to update tests might be OK. Especially since [`ThreadId`](https://doc.rust-lang.org/std/thread/struct.ThreadId.html), for some inexplicable reason, has no stable way to get the actual ID (only in nightly).

Reply:
This is actually just some debugging remnant :see_no_evil: but I think it makes sense to assert the panic message to verify.